### PR TITLE
[canvaskit] Only check missing glyphs against fallback fonts once per frame

### PR DESCRIPTION
**Pull request by hterkelsen**
Calling `getGlyphIDs` is very slow (especially for the large emoji font) since CanvasKit decodes the font for every check. This causes frame drops when there is a lot of separate text that needs to be checked in a frame.

With this PR, we only check `getGlyphIDs` for the fallback fonts once per frame, which largely avoids the slowdown.

Fixes https://github.com/flutter/flutter/issues/79882

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/text.dart [hunk 1]
```diff
@@ -902,176 +905,21 @@ class CkParagraphBuilder implements ui.ParagraphBuilder {
...
-    final FontFallbackData fallbackData = FontFallbackData.instance;
-    codeUnits.removeWhere((int codeUnit) =>
-        fallbackData.codeUnitsWithNoKnownFont.contains(codeUnit));
-    return codeUnits.isEmpty;
-  }
-
-  @override
-  void addText(String text) {
-    _ensureFontsSupportText(text);
+    FontFallbackData.instance.ensureFontsSupportText(text, fontFamilies);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/text.dart:918]: It's great that all the font-related stuff is moving out of the paragraph code. Separation of concerns FTW!

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart [hunk 2]
```diff
@@ -70,6 +70,164 @@ class FontFallbackData {
...
+  bool _scheduledCodeUnitCheck = false;
+
+  /// Determines if the given [text] contains any code points which are not
+  /// supported by the current set of fonts.
+  void ensureFontsSupportText(String text, List<String> fontFamilies) {
+    // TODO(hterkelsen): Make this faster for the common case where the text
+    // is supported by the given fonts.
+
+    // A list of unique code units in the text.
+    final List<int> codeUnits = text.runes.toSet().toList();
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart]: Done

File: lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart [hunk 3]
```diff
@@ -70,6 +70,164 @@ class FontFallbackData {
...
+    if (isAscii) {
+      return;
+    }
+
+    // First, check if every code unit in the text is known to be covered by one
+    // of our global fallback fonts. We cache the set of code units covered by
+    // the global fallback fonts since this set is growing monotonically over
+    // the lifetime of the app.
+    codeUnits.removeWhere(
+        (int codeUnit) => knownCoveredCodeUnits.contains(codeUnit));
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart]: Good suggestion! Done

**Pull request merged**